### ISSUE TITLE
Allow list in workflow

### ISSUE THREAD

**Issue by GregHydeDartmouth** [RTXteam/RTX#1926]
Hi all, I have a very general question on allow lists for ARAX workflows. Additionally, I have a potential bug I've surfaced in working with allow lists in ARAX.

Is there a way to use an allowlist that includes all teams without having list them all? I'm working on a 3 hop that goes from Disease-e0->Tissue-e1->Gene-e2->Drug. I specified an allowlist for e1 using CHPs Tissue-Gene specificity edge as follows:
```
{
    "id": "fill",
    "parameters": {
        "allowlist": [
            "infores:connections-hypothesis"
        ],
        "qedge_keys": [
            "e1"
        ]
    }
}
```
However when I do this, ARAX seems to want me to fill in an allow list for all other edges, but I don't want to constrain the other hops to any particular team. For this reason I specify the other edges as follows:
```
{
    "id": "fill",
    "parameters": {
        "allowlist": [
            'infores:automat-cord19', 'infores:text-mining-provider-cooccurrence', 'infores:automat-uberongraph', 'infores:automat-robokop', 'infores:automat-covidkop', 'infores:automat-hetio', 'infores:cam-kp', 'infores:service-provider-trapi', 'infores:rtx-kg2'
        ],
        "qedge_keys": [
            "e0"
        ]
    }
}
```
My question is, am I able to specify an allowlist that includes all teams without listing each name? I've tried nulling allowlist, providing an empty list, and removing the allowlist parameter all together to no avail. 

I also noticed potential erroneous behavior when listing each team member explicitly in the allowlist.
1.) When I specify the teams explicitly, the edge is passed along to those teams 1 at a time, rather than in a batch as I would expect. For example, it sends e0 to infores:automat-cord19, waits for it to respond and then sends e0 to infores:text-mining-provider-cooccurence, and on. This makes the querying a little bit slower. 
2.) When I specify an allow list explicitly, it appears to reduce the curies passed along from each hop to the curies that can be answered by ALL teams, rather than custom tailoring the curies to each team according to their meta_kg expected curie formats. 

The full query I'm trying to answer is here:
```
{
    "workflow": [
        {
            "id": "fill",
            "parameters": {
                "allowlist": [
                    #'infores:automat-cord19', 'infores:text-mining-provider-cooccurrence', 'infores:automat-uberongraph', 'infores:automat-robokop', 'infores:automat-covidkop', 'infores:automat-hetio', 'infores:cam-kp', 'infores:service-provider-trapi', 'infores:rtx-kg2'
                    'infores:rtx-kg2'
                ],
                "qedge_keys": [
                    "e0"
                ]
            }
        },
        {
            "id": "fill",
            "parameters": {
                "allowlist": [
                    "infores:connections-hypothesis"
                ],
                "qedge_keys": [
                    "e1"
                ]
            }
        },
        {
            "id": "fill",
            "parameters": {
                "allowlist": [
                    #'infores:automat-drug-central', 'infores:automat-covidkop', 'infores:automat-gtopdb', 'infores:automat-pharos', 'infores:sri-reference-kg', 'infores:automat-hmdb', 'infores:automat-robokop', 'infores:molepro', 'infores:rtx-kg2', 'infores:service-provider-trapi', 'infores:automat-hetio', 'infores:cam-kp', 'infores:automat-ctd', 'infores:rtx-kg2'
                    'infores:rtx-kg2'
                ],
                "qedge_keys": [
                    "e2"
                ]
            }
        },
        {
        "id": "bind"
        },
        {
        "id": "score"
        }
    ],
    "message": {
        "query_graph": {
            "nodes": {
                "n0": {
                    "categories": [
                        "biolink:Disease"
                    ],
                    "ids": [
                        "MONDO:0009061"
                    ]
                },
                "n1": {
                    "categories": [
                        "biolink:GrossAnatomicalStructure"
                    ]
                },
                "n2": {
                    "categories": [
                        "biolink:Gene"
                    ]
                },
                "n3": {
                    "categories": [
                        "biolink:Drug",
                        "biolink:SmallMolecule"
                    ]
                }
            },
            "edges": {
                "e0": {
                    "subject": "n0",
                    "object": "n1",
                    "predicates": [
                        "biolink:located_in"
                    ]
                },
                "e1": {
                    "subject": "n1",
                    "object": "n2",
                    "predicates": [
                        "biolink:expresses"
                    ]
                },
                "e2": {
                    "subject": "n3",
                    "object": "n2",
                    "predicates": [
                        "biolink:affects"
                    ]
                }
            }
        }
    }
}
```
Note I've commented out the long list of infores names on e0 and e2. I alternated between using only rtx-kg2 and the long list by commenting one in, and the other out. 

Here is an example of using a long explicit allowlist on e0 and e2: https://arax.ci.transltr.io/?r=72150. Here chp returns 170 edges for e1, but only 7 uniprot ids get passed along in e2 to all teams.
Here is an example of using rtx-kg2 on e0 and e2 https://arax.ci.transltr.io/?r=72149. Here chp returns 180 edges for e1 and rtx receives a broad range of identifiers from uniprot, ncbi, ensembl, fma, etc. 
It's possible there is another explanation for what I'm seeing, but it's hard to tell strictly from observing the responses and QGs sent to teams.

**Comment by amykglen** [comment]
Hi @GregHydeDartmouth - thanks for the detailed report. Indeed this sounds like strange behavior.

I think this is probably due to bug(s) in our `operation_to_ARAXi.py` and maybe some other modules.

@dkoslicki - since this used to be Finn's module do we have a plan for who should work on it now? For instance, probably these two lines:
https://github.com/RTXteam/RTX/blob/0ec5c8f57c3272dbad56d64f7163cdc7b9e97245/code/ARAX/ARAXQuery/operation_to_ARAXi.py#L126-L127
should be changed to something like:
```
      ARAXi.append(_make_expand_query(parameters['allowlist'], parameters.get('qedge_keys')))
```
(so that Expand isn't called individually per KP)

I don't think ARAXi yet allows you to specify multiple KPs for one `expand()` command, but it probably should. (the adjustment for that internally within Expand should be pretty minor since Expand already operates in a world where it expects to have a list of KPs per qedge.)

**Comment by dkoslicki** [comment]
@amykglen When you get back, does this mean this is blocking on `expand` allowing for multiple KPs to be specified? I'm happy to maintain `operation_to_ARAXi.py`, just need to know how to change the call to `expand` in ARAXi.

@GregHydeDartmouth the behavior of `allow_list` "should" have the behavior that if you don't include it, every KP that advertises it can respond to the query is actually queried. I'll let you know when I have a solution.

**Comment by amykglen** [comment]
that's right, @dkoslicki - and I just updated Expand to allow lists of KPs in ARAXi (in `master`) - the syntax for doing so looks like this:
```
expand(kp=[infores:rtx-kg2, infores:molepro])
```

so I think the tweak to `operation_to_ARAXi.py` will need to be like what I described in this comment: https://github.com/RTXteam/RTX/issues/1926#issuecomment-1293730108

(basically you want to make one `expand()` call with _all_ KPs in the allowlist, rather than separate `expand()` calls for each KP)

I think this should solve issues 1) and 2) Greg described (i.e., KPs not being queried in parallel and KPs' answers being limited to other KPs' answers), but I'm not sure why ARAX seems to be requiring him to fill in an allowlist for _all_ qedges when he provides it for one. I think that issue may be on the `operation_to_ARAXi.py` side? if you provide no `kp=` parameter when calling `expand()`, it defaults to using all KPs capable of answering that query.
